File: Nmap-Detector/ND.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/ip.h>
#include <netinet/tcp.h>
#include <arpa/inet.h>
#include <sys/wait.h>

void trigger_tshark_capture(int capture_count) {
    pid_t pid = fork();

    if (pid == 0) {

        char filename[50];
        snprintf(filename, sizeof(filename), "captures_nmap/cnmap%d.pcap", capture_count);

        char *args[] = {"tshark", "-i", "any", "-a", "duration:600", "-w", filename, NULL};
        execvp("tshark", args);

        perror("Failed to run Tshark");
        exit(EXIT_FAILURE);
    } else if (pid < 0) {

        perror("Fork failed");
        exit(EXIT_FAILURE);
    } else {

        wait(NULL);
    }
}

int detect_nmap_scan(unsigned char *buffer, int data_size) {
    struct iphdr *iph = (struct iphdr *)buffer;

        if (iph->protocol == 6) {
            struct tcphdr *tcph = (struct tcphdr *)(buffer + iph->ihl * 4);

            if (tcph->syn == 1 && tcph->ack == 0) {
                printf("Potential Nmap SYN scan detected!\n");
                return 1;
        }
        }

    return 0;
}

int main() {
    int sock_raw;
    unsigned char *buffer = (unsigned char *)malloc(65536);
    struct sockaddr saddr;
    int saddr_size = sizeof(saddr);
    int data_size;
    int capture_count = 1;

    sock_raw = socket(AF_INET, SOCK_RAW, IPPROTO_TCP);
    if (sock_raw < 0) {
        perror("Socket Error");
        return 1;
    }

    printf("Listening for packets...\n");

    while (1) {

        data_size = recvfrom(sock_raw, buffer, 65536, 0, &saddr, &saddr_size);
        if (data_size < 0) {
            perror("Recvfrom error, failed to get packets");
            return 1;
        }

        if (detect_nmap_scan(buffer, data_size)) {

            trigger_tshark_capture(capture_count);
            capture_count++;
            sleep(600);
        }
    }

    close(sock_raw);
    free(buffer);
    return 0;
}

File: Nmap-Detector/ND.py
#This code is written by chatgpt for those who don't want to use C program
#Feel free to modify

import os
import socket
import struct
import time
import subprocess

def trigger_tshark_capture(capture_count):
    """
    Trigger tshark to capture network traffic for 10 minutes and save it in captures_nmap directory.
    """
    filename = f"captures_nmap/pnmap{capture_count}.pcap"
    command = ["sudo", "tshark", "-i", "any", "-a", "duration:600", "-w", filename]

    # Run tshark as a subprocess
    subprocess.Popen(command)
    print(f"Started capturing traffic into: {filename}")

def detect_nmap_scan(packet):
    """
    Detects potential Nmap SYN scan by inspecting the TCP SYN flag.
    """
    ip_header = packet[0:20]
    iph = struct.unpack('!BBHHHBBH4s4s', ip_header)

    protocol = iph[6]
    if protocol == 6:  # Check if the protocol is TCP (6)
        tcp_header = packet[20:40]
        tcph = struct.unpack('!HHLLBBHHH', tcp_header)

        syn_flag = tcph[5] >> 1 & 1
        ack_flag = tcph[5] >> 4 & 1

        if syn_flag == 1 and ack_flag == 0:  # SYN set, ACK not set
            print("Potential Nmap SYN scan detected!")
            return True
    return False

def main():
    try:
        # Create a raw socket to listen for TCP packets
        sock = socket.socket(socket.AF_INET, socket.SOCK_RAW, socket.IPPROTO_TCP)
    except socket.error as msg:
        print(f"Socket could not be created. Error Code: {str(msg)}")
        return

    # Ensure the captures_nmap directory exists
    if not os.path.exists('captures_nmap'):
        os.makedirs('captures_nmap')

    capture_count = 1
    print("Listening for packets...")

    while True:
        # Receive packets
        packet = sock.recvfrom(65565)[0]

        # Detect Nmap scan
        if detect_nmap_scan(packet):
            trigger_tshark_capture(capture_count)
            capture_count += 1

            # Sleep for 10 minutes to avoid triggering multiple captures
            time.sleep(600)

if __name__ == "__main__":
    main()

